gtn/functions/compose.cpp

```diff
@@ -450,3 +456,936 @@ Graph compose(
 
 } // namespace detail
 } // namespace gtn
+
+namespace gtn {
+namespace detail {
+namespace dataparallel {
+// Change AOS to SOA
+struct GraphDataParallel {
+  // True if a node is accept or start, false otherwise
+  std::vector<bool> accept;
+  std::vector<bool> start;
+
+  // One value per node - i-th value corresponds to i-th node
+  // Last element is the total number of arcs, so that
+  // each element and its neighbor forms a range
+  std::vector<int> inArcOffset;
+  std::vector<int> outArcOffset;
+
+  // One value per arc
+  std::vector<int> inArcs;
+  std::vector<int> outArcs;
+
+  // One value per arc
+  // i-th value corresponds to i-th arc
+  std::vector<int> ilabels;
+  std::vector<int> olabels;
+  std::vector<int> srcNodes;
+  std::vector<int> dstNodes;
+  std::vector<float> weights;
```

Comment:
Just leaving this as a not for the future: these are stored in the same order as the original graph's arcs right? So there may be some opportunity to not duplicate the full graph. (I can see how they are convenient to have for now).

Reply:
Yeah some of it is convenience for now for easy debugging etc :)